collections/_article/whats-new-in-csharp-for-godot-4-0.md

```diff
@@ -0,0 +1,357 @@
+
+#### Relevant links
+
+- Make GodotCompletionProviders work via NuGet ([GH-18](https://github.com/godotengine/godot-csharp-visualstudio/issues/18)).
+
+## Future
+
+Some of the features that didn't make it in time for Godot 4.0 and that we already plan to work on:
+
+### Global classes
+
+Global classes, also known as named scripts, are classes registered in the editor so they can be used more conveniently. These classes appear in the _Add Node_ and _Create Resource_ dialogs. GDScript supports this feature with the `class_name` syntax.
+
+We tried, but ultimately failed, to bring this feature to C# in time for 4.0, but there's already a PR open so it's very likely it will be ready for the next release, Godot 4.1.
+
+#### Relevant links
+
+- Add first-class custom resource support ([GH-18](https://github.com/godotengine/godot-proposals/issues/18)).
+- Allow exporting custom resources from/to GDScript, VisualScript, C#, and PluginScript ([GH-48201](https://github.com/godotengine/godot/pull/48201)).
+- Script-class-aware Inspector & related controls ([GH-62413](https://github.com/godotengine/godot/pull/62413)).
+- Enable QuickOpen to see scripted resources ([GH-62417](https://github.com/godotengine/godot/pull/62417)).
+- Add C# resource export ([GH-72619](https://github.com/godotengine/godot/pull/72619)).
+
+### Mobile and web support
+
+We aim to support C# in all the platforms that Godot is available on. Unfortunately, we were unable to implement support for mobile and web platforms for 4.0, these platforms gained support upstream somewhat recently so we didn't have much time to work on it. The new workloads in .NET 6 should allow us to support the mobile and web platforms really soon, so keep an eye out for that.
+
+#### Relevant links
+
+- [Announcing .NET 6 — The Fastest .NET Yet](https://devblogs.microsoft.com/dotnet/announcing-net-6/).
+
+### Full AOT and trimming support
+
+This can greatly reduce binary size and enable further performance optimizations.
+
+In order to support this, we'll need to ensure that our libraries are prepared. Dynamically accessed types can't be trimmed, so we need to avoid reflection as much as possible. Some platforms don't support JIT, so we definitely need to support AOT. We'll be working on this as we add support for more platforms in future 4.x releases.
+
+#### Relevant links
+
+- [App Trimming in .NET 5](https://devblogs.microsoft.com/dotnet/app-trimming-in-net-5/).
+- [Native AOT deployment](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/).
+
+### GDExtension support
+
+The ability to create GDExtensions using C# would enable powerful workflows.
+
+This would allow us to avoid the limitations of a scripting language implementation, such as [relying on file paths](https://github.com/godotengine/godot/issues/15661) to reference C# classes. Users will be able to implement C# types that can be registered in ClassDB and behave as built-in Nodes and Resources.
```

Comment:
```suggestion
The ability to create GDExtensions using C# will enable new and powerful workflows. This would allow us to avoid the limitations of a scripting language implementation, such as [relying on file paths](https://github.com/godotengine/godot/issues/15661) to reference C# classes. Users are going to be able to implement C# types that can be registered in Godot's internal type database, ClassDB, and that would behave as built-in Nodes and Resources.
```